file_catalog/server.py

```diff
@@ -686,6 +687,101 @@ async def put(self, uuid: str) -> None:
+    """Initialize a non-RESTful action handler for removing an existing record's location.
+
+    And potentially the entire record.
+    """
+
+    def initialize(self, **kwargs: Any) -> None:  # type: ignore[override]  # pylint: disable=C0116,W0221
+        """Initialize handler."""
+        super().initialize(**kwargs)
+        # pylint: disable=W0201
+        self.files_url = os.path.join(self.base_url, 'files')
+
+    @validate_auth
+    @catch_error
+    async def post(self, uuid: str) -> None:
+        """Handle POST request.
+
+        Remove location from the record identified by the provided UUID,
+        and potentially the entire record.
+        """
+        # try to load the record from the file catalog by UUID
+        try:
+            db_file = await self.db.get_file({'uuid': uuid})
+        except pymongo.errors.InvalidId:
+            self.send_error(400, reason='Not a valid uuid')
+            return
+        if not db_file:
+            self.send_error(404, reason='File uuid not found')
+            return
+
+        # decode the JSON provided in the POST body
+        body = json_decode(self.request.body)
+        try:
+            site = body.pop("site")
+            path = body.pop("path")
+        except KeyError:
+            self.send_error(400, reason="POST body requires 'site' & 'path' fields")
+            return
+        if body:
+            # REASONING:
+            # If client defines (site=X, path=Y, archive=True)
+            # - does this match (site=X, path=Y)?
+            # - or (site=X, path=Y, archive=False)?
+            # - or only (site=X, path=Y, archive=True)?
+            # What if they don't define archive at all? (site=X, path=Y)
+            # - does this match (site=X, path=Y, archive=True)?
+            # It's unclear, so better fail fast
+            self.send_error(
+                400,
+                reason=f"Extra POST body fields detected: {list(body.keys())} "
+                       f"('site' & 'path' are required)"
+            )
+            return
+
+        def is_location_match(loc: types.LocationEntry) -> bool:
+            # only match against the mandatory fields
+            return bool(loc['site'] == site and loc['path'] == path)
+
+        # Remove `location` (possibly entire record) & Send Back
+        before = db_file.get('locations', [])
+        after = [loc for loc in before if not is_location_match(loc)]
+        # bad location! -- no location was filtered out
+        if before == after:
+            self.send_error(
+                404,
+                reason=f"Location entry not found for site='{site}' & path='{path}'"
+            )
+            return
+        # remove location! -- there are remaining locations after filtering
+        elif after:
+            await self.db.update_file(uuid, {'locations': after})
+            # re-read the updated record from the database
+            db_file = await self.db.get_file({'uuid': uuid})
```

Comment:
Consider either modifying or creating a new db method that does `find_one_and_update` so it can atomically update the file and return it.